### ISSUE TITLE
Including our app as a project in a FOSS internship program

### ISSUE THREAD

**Issue by misaochan** [commons-app/apps-android-commons#815]
I figure we could potentially put our app up for the next [Outreachy](https://www.gnome.org/outreachy/) round (Dec - Mar). AFAIK GSoC is only once a year and we have missed the last intake, so Outreachy will be the next one up. This is always a good way to get more visibility for us and to introduce new programmers to our project (plus the benefit of Outreachy is that it is for underrepresented groups in FOSS, hence more diversity).

If we decide to do this, I can likely be one of the mentors for the project (can't confirm just yet, but most likely I will have the time to do so at the end of the year). Wikimedia projects usually require two mentors though, so even if I can do it, we will need someone else to be a co-mentor. Anyone here interested? :)

**Comment by psh** [comment]
What is involved (time commitment, or other prerequisites) in being a mentor?

**Comment by misaochan** [comment]
It varies depending on how involved you want to be, as far as I can tell. I would estimate 5 hours per week or so. Perhaps @nicolas-raoul can provide better information on this. ;)

**Comment by neslihanturan** [comment]
As an old intern of Outreachy program, I would be very happy to mentor it to increase diversity.

**Comment by misaochan** [comment]
Thanks @neslihanturan ! Once I can confirm my availability, at the end of August maybe we can take a look around the issues board and pick a few suitable tasks? Even if I end up not being available, if anyone else is willing to be the 2nd mentor then we can still go ahead. :)

**Comment by misaochan** [comment]
Hi all,

Unfortunately, I do not think I can be the primary mentor for this round. However, I am happy to be the secondary mentor/co-mentor, if someone else would like to be the primary. The primary mentor would be the main point of contact for the intern as well as for the Wikimedia and Outreachy coordinators, and would be responsible for reviewing their code. I will help and advise wherever I can, participate in the evaluations and attend any meetings that I am able to.

@neslihanturan are you interested in being the primary mentor? No pressure, if you don't think you can afford the time then that's totally fine, we can skip this round. :) If you or anyone else is interested, we can go ahead with this and start looking for a suitable project and post on the Phab board for this round when that is up.

**Comment by sandarumk** [comment]
What about Google Code In? Unlike other mentorship projects, the tasks for GCI are particularly smaller since they are targeted to high school students. The tasks do not have to be only coding, they can be testing, documentation, outreach, marketing, etc. Tasks should have designated mentors, but one task can have many mentors, so the workload can be shared.

**Comment by neslihanturan** [comment]
Actually I think affording mentoring time is not a big deal for me, however I don't feel qualified enough to be primary mentor, yet. If @psh can have time to be primary mentor as a more experienced developer, I would be very happy to be co-mentor. 
On the other hand, I am not sure about a suitable task that is big enough to be held in 3 months. Do you have any idea?

Regarding to Google Code In issue, yes we can also consider it. However my willingness  is only valid for FOSS program. We can still look for new mentors.

**Comment by misaochan** [comment]
@neslihanturan Sure, if @psh is able to participate, I am happy for both of you to be the mentors (and you guys can decide between yourselves who will be the primary or secondary mentor :) ). In that case I'll grant him collaborator  permissions (which he will need for that), and remember to ping me to post on the Phab thread, as AFAIK there used to be a rule that said that projects will need the project maintainers approval.

If I recall correctly, the guideline for the task chosen is that it should take an experienced developer a month or less to complete, and there should be a minimum viable product ready by the midpoint report. We can take a look at the issues we have later on and see which ones will fit. 

@sandarumk Yeah, GCI is a viable option too. :) There is no reason why we can't participate in both, if enough people are willing to be mentors. However I am not as familiar with the GCI process as I am with the Outreachy process.

**Comment by psh** [comment]
Yes, I'd be happy to be a mentor.  @neslihanturan - you've been around this project longer than me and you know the Outreachy better than I do - I think its significant that you've been through the program yourself.  You're uniquely qualified to take a lead role!

Issue #725 might be a candidate for a project - incorporating tabs into the main activity of the app, adding a FAB and converting the existing `ListView` over to a `RecyclerView`.  There are a bunch of smaller steps - 

1. Remove the contributions list menu and replace it with a FAB that shows/hides additional options
2. Pull existing `NearbyActivity` GUI into a Fragment & host it on the `NearbyActivity` 
3. Create a new "home activity" with tabs (`ContributionsListFragment` and the new `NearbyFragment`)
4. (If there is time) convert the `ContributionsListFragment` to using a `RecyclerView` (last remaining item from Issue #743)

Each step has a defined end point that would be releasable.  Depending on how well they do, an MVP might be 1+2, or 1+2+3.

**Comment by neslihanturan** [comment]
I also think #725 is suitable, however, improving our UI is one of our goals from our proposal. So, it is also important for us to make it happen in 6 months. We might need to finish this task before Outreachy session finishes. So maybe we can think of a new features or enhancements that are not written in our proposal.

For example there was an idea about gamifying the map here #334. Can we consider this as Outreachy proejct, or would it be too hard?

**Comment by misaochan** [comment]
Thanks for volunteering, @psh ! :) So it's settled - you and Neslihan can be the mentors. I would strongly recommend reading https://www.mediawiki.org/wiki/Outreachy/Mentors and the associated links beforehand.

I agree with @neslihanturan - it's too late for us to make significant changes to the IEG renewal proposal, plus honestly getting an intern is not a guarantee. In most rounds, some projects don't move forward with interns, usually because none of the interns who applied for it were accepted by Outreachy. So unfortunately #725 is not an option. 

"Gamifying the map" can be a potential task but needs a lot of further fleshing out. E.g. what sort of gamification, etc. Then we still need to break it down to smaller tasks. Plus there is an additional concern that we will likely be working on the Nearby map significantly during Dec/Jan, so the intern will encounter a lot of confusion with rapidly-changing code and conflict resolution, which might be a bit much for a beginner to cope with.

Personally my suggestion would be very specific, encapsulated, and easy-to-scope tasks like #161 + #844 , or just #781 perhaps?

**Comment by misaochan** [comment]
Srishi Sethi (of WMF) recently posted on my talk page about this, along with instructions for how to include our project. Please feel free to chime in in the conversation or ask her further questions if needed. Link: https://www.mediawiki.org/w/index.php?title=Topic:Tx36gt6b4ppmrzto&topic_showPostId=tx36gt6cou5a46jg&fromnotif=1#flow-post-tx36gt6cou5a46jg

**Comment by srish** [comment]
Glad to hear there is an interest in including Commons App project in Outreachy! :) @misaochan has already included all the relevant resources. Application period opens in about a week, so if you could let me know which project you would want to showcase, that would be great. This time we are planning to promote the featured projects during the application period, can't guarantee but we'll try our best to help recruit potential candidates.

**Comment by srish** [comment]
Checking if there is an interest in showcasing the project for this round  https://www.mediawiki.org/wiki/Outreachy/Round_15 :) Let me know soon!

**Comment by misaochan** [comment]
@neslihanturan , are you around? I think you and @psh need to decide on a task and start the process of including our project (see @srish 's comments). Admittedly this might be somewhat difficult timing, with the IEG results coming up soon and a lot of uncertainty and busyness in the meantime, so if you guys want to postpone our involvement to a future round it's totally understandable.

If you two want to go ahead with this round, I am good with any task that you two decide on as long as it isn't in the current IEG renewal proposal. If you really want to do #334 and feel that you can come up with a concrete outline for it in time, by all means go for it. :)

**Comment by neslihanturan** [comment]
No, I am still on the vacation, this is the reason of my silence. Sorry for this. Actually IEG results uncertainty makes taking responsibility decision hard for me. I think postponing it to summer round may be a better idea?

**Comment by neslihanturan** [comment]
No, I am still on the vacation, this is the reason of my silence. Sorry for this. Actually IEG results uncertainty makes taking responsibility decision hard for me. I think postponing it to summer round may be a better idea?

**Comment by misaochan** [comment]
Yeah, postponing sounds good to me in that case.

**Comment by srish** [comment]
Hello folks! Checking here again to see if there is an interest in mentoring this project in the Google Summer of Code'18 or Outreachy'16? I am asking as we are currently recruiting projects for both programs.. they are coming up :) https://www.mediawiki.org/wiki/Google_Summer_of_Code/2018
https://www.mediawiki.org/wiki/Outreachy/Round_16

**Comment by nicolas-raoul** [comment]
@maskaravivek

**Comment by maskaravivek** [comment]
@misaochan Since past few days, i was talking to @nicolas-raoul about mentoring programs and have started as a co-mentor in few of the Google code in tasks. I am up for mentoring for GsoC or Outreachy as a mentor/co-mentor. Are we still planning to submit our project for this round.

**Comment by misaochan** [comment]
@maskaravivek Great! Please feel free to submit our project for this round if you are willing to be a mentor (perhaps ask @psh or @neslihanturan if they would be willing to be your co-mentor?). You can look through our issues to see which ones will be suitable for it. 

Unfortunately I am personally swamped this year and cannot help, but I don't think I actually need to do anything other than post on the Phab task stating that I support the planned improvements as the project maintainer. So you guys can absolutely go right ahead. :)

**Comment by neslihanturan** [comment]
I am also very busy this year, even if I want to be volunteer for co-mentoring, have to stop myself and focus other tasks of mine:/

**Comment by nicolas-raoul** [comment]
Tips to everyone willing to so a GSoC'18 or Outreachy'16:

1. [Here](https://phabricator.wikimedia.org/T115101) is a past application that could can use as a model if you want. That particular application was accepted and successful, actually the person who wrote it eventually became the project's maintainer.
2. Feel free to post your proposition at https://github.com/commons-app/apps-android-commons/issues/new so that we can provide feedback on it and give you more tips.
3. Below are my suggestions for a GSoC with an achievable goal, that would truly make the app better, easy enough to explain on a resume, and not too hard nor overly easy:

A: Browse pictures and see uploaded / not uploaded status #175

B: Option to "anonymize" pictures #70 + Remove from EXIF: geo info if too close to home #181

C: Show user talk notifications #7 + Feedback on how my pictures get used: Statistics #85 (statistics only) + Tutorial again and quizz, for people who did not understand topicness/copyright immediately #274 (as far as time allows)

D: Nearby notable things from OpenStreetMap with no Wikidata item #883

E: Peer review - review the latest pictures uploaded by other people #780

F: Localized category search #971 + Apply the user's second preferred language to place labels #516 + Nearby item label: If user's languages not available, show in the item's language if available #729 (if time allows)

**Comment by maskaravivek** [comment]
@misaochan @neslihanturan Yes, I understand you guys have a packed schedule. :). Just to understand clearly, what are the steps needed from our side to include our project for GSoC/Outreacy. I can sift through the suggested tasks above and then come up with filtered options that would both benefit our project and would be doable within a 2-3 month period. 

@nicolas-raoul Thanks for the tip. 

@psh Would you be able to spare some time from your schedule and help in mentoring?

**Comment by misaochan** [comment]
@maskaravivek I think https://www.mediawiki.org/wiki/Outreachy/Mentors has a good overview of the steps needed. But I have never been a mentor personally so not sure if that alone is enough, maybe @srish can point out other documentation needed?

**Comment by misaochan** [comment]
Also, the list @nicolas-raoul provided is great, with the exception of "C: Show user talk notifications #7 " - that is planned for our IEG and I believe work has already started on it.

**Comment by nicolas-raoul** [comment]
@misaochan Oh sorry I removed that point, I believe that even without it C still has enough substance. Thanks :-)

**Comment by misaochan** [comment]
I agree! C is already very substantial with just the 2 tasks IMO. :)

**Comment by maskaravivek** [comment]
I took a look at the possible tasks suggested by @nicolas-raoul @misaochan and other open issues. I won't be adding any new choices here. :D Here are a few tasks from the above lists that I feel would have sufficient scope for a 2-3 month internship and would help our app evolve. I would love to mentor students with any of these tasks.

- Feedback on how my pictures get used: Statistics #85 (statistics only) + Tutorial again and quizz, for people who did not understand topicness/copyright immediately #274 (as far as time allows)
- Peer review - review the latest pictures uploaded by other people #780

I have gone through the responsibilities of a mentor and would be able to meet the time commitments expected by a mentor. I have also added myself to the list of available mentors. 

@srish Could you guide us through anything else that is needed from us to make our project available for GSoC/Outreachy programs.

**Comment by srish** [comment]
@maskaravivek So glad to hear that you are onboard for this! Next steps are that you create a task on Phabricator for these projects, and tag them with #Google-Summer-of-Code (2018) and/or #Outreachy (Round-16). You can see some task descriptions for projects in https://phabricator.wikimedia.org/project/view/3096/ for reference. I will then showcase them on https://www.mediawiki.org/wiki/Google_Summer_of_Code/2018 and https://www.mediawiki.org/wiki/Outreachy/Round_16. We could then also try to recruit a co-mentor for you from mailing lists, and elsewhere.

**Comment by misaochan** [comment]
@maskaravivek Great job! Thanks for offering to mentor. :) 

A few things:
- We will probably need to flesh out the scope of #274 a little beforehand. Currently the requirements are still a bit vague. I would **not** be in favour of locking the app for users who fail the quiz, also. :) Let us discuss things further on that issue
- For the "skills required", do we really need familiarity with Kotlin? AFAIK Kotlin is only used for a few of our tests and is completely optional. I worry if stating that might scare away potential applicants. Maybe make it an "optional but not required" thing?

**Comment by misaochan** [comment]
Just talked to @maskaravivek about this - if an active co-mentor can't be found by the time the applications begin, I can lend my support as co-mentor. :) I've told him that I may not have the time to do much so the majority of the workload might fall on him, but if we can't get someone with more availability, I'm happy to sign up so we can move the project ahead.

**Comment by misaochan** [comment]
Hi @srish ,

We have been receiving a huge number of contributions from potential applicants. :) In view of that, @neslihanturan has volunteered to join @maskaravivek and I in mentoring if we are allowed to have 2 interns in our project, in order to give as many applicants as possible the chance to participate in GSoC/Outreachy. Would that be at all possible? We are also talking about a potential 2nd topic for the 2nd intern, if allowed.

Totally understand if budget or organization rules prevent this, though.

**Comment by srish** [comment]
@misaochan that sounds wonderful :) Thanks to you and @neslihanturan for willing to mentor as well! :) It is totally possible that we accept more than one intern for your project and promote another topic. The only thing to keep in mind is that for Outreachy we have agreed to fund up to two interns. So, if we are to select a second intern; we could do so only through the Google Summer of Code slots (for which there is no such limit). Let me know if you've questions!

**Comment by misaochan** [comment]
Thanks @srish ! If we do that, should we create a new Phab task for the 2nd topic? Or should we have both under the same task?

**Comment by srish** [comment]
@misaochan Yes, it would be good to create a Phab task.

**Comment by nicolas-raoul** [comment]
Nice to hear that GSoC is getting so much attention this year, and that everyone is stepping up to help! :-)
@neslihanturan: Have you decided the task already? I guess you are looking for a co-mentor? @whym @VojtechDostal @psh @domdomegg @akaita @veyndan would you be interested?

**Comment by misaochan** [comment]
I would suggest #1221 for the second topic , if you guys like the sound of that.

**Comment by misaochan** [comment]
@srish If we can't get a co-mentor for the 2nd task, is it okay to just list myself or @maskaravivek as co-mentor even though we are already on the 1st task? Or would it be best to wait for a 4th volunteer?

**Comment by 97balakrishnan** [comment]
I think #697 can be added here.   ?

**Comment by nicolas-raoul** [comment]
If finding a co-mentor for the second GSoC proves too difficult, I can be co-mentor.

**Comment by misaochan** [comment]
Wonderful, thanks @nicolas-raoul ! 👍 

@97balakrishnan , I think we already have 2 tasks? I doubt we can manage a 3rd. ;)

**Comment by 97balakrishnan** [comment]
@misaochan Yeah ok :+1: thanks. Pls create a Phab task. :)

**Comment by misaochan** [comment]
Phab task is up! https://phabricator.wikimedia.org/T188877

**Comment by 97balakrishnan** [comment]
Thanks a lot :)

**Comment by misaochan** [comment]
So, looking through the GSoC and Outreachy guidelines for this round, it seems that the application period for [Outreachy](https://www.outreachy.org/apply/project-selection/) is anytime til March 22, and for [GSoC](https://summerofcode.withgoogle.com/how-it-works/#timeline) is March 13 - 28? I guess this means that we will need to get everything sorted for the students to start applying by March 13 latest, right?

**Registration:** 
@maskaravivek and I have registered our project for Outreachy and signed up as mentors. I guess @neslihanturan and @nicolas-raoul will need to do the same? Granted we can only have 1 Outreachy intern max (and the other must come from GSoC), but we don't know which project will get which, so probably best to register both?

How does registration work for GSoC, does anyone know?

**Fleshing out scope of tasks:**

I think the Phab task of project 2 (#1221 ) is mostly ready, just needs a bit of tweaking. I will add the proposed schedule for project 1 (#274 and #85 ) to the Phab task once we get a bit more feedback on those issues.

**Other**
Anything else that I missed?

**Comment by srish** [comment]
@misaochan hello! Thanks for taking the lead in pulling this off, also @neslihanturan and @nicolas-raoul :) 

As we have already showcased a commons-app project for Outreachy and our budget would allow us to only fund up to two candidates, I propose we promote this project only in Google Summer of Code. But, if you have other suggestions let me know. 

If we end up promoting this project only in GSoC, you don't have to do anything right now. I will add your new project on MW GSoC page and share among students in Zulip chat. In the next week or so, you will be supposed to register on the GSoC program website. For that, I will send you all an invite from the GSoC dashboard on your preferred email contact in next few days or so (drop me a note at ssethi@wikimedia.org and let me know which email Id you would like to use for this purpose).

The process for getting feedback on the proposals is still the same. https://www.mediawiki.org/wiki/Google_Summer_of_Code/Participants#Application_process_steps (Step 9).

**Comment by misaochan** [comment]
Thanks so much @srish ! Will email you. :)

**Comment by misaochan** [comment]
Hi @srish ,

Just confirming that Outreachy participants will also need to create a proposal on Phab? I checked phabricator.wikimedia.org/project/view/3097/ but it seems that the workboard hasn't been created yet?

**Comment by srish** [comment]
@misaochan Yes! good point. The workboard lives here now > https://phabricator.wikimedia.org/project/board/3097/. Outreachy candidates can now add their proposals on Phabricator.

**Comment by srish** [comment]
Hi all! Just checking if we are showcasing the following project: https://phabricator.wikimedia.org/T188877 only for GSoC 2018, right?

**Comment by misaochan** [comment]
I think so, based on our discussion above. I guess the Outreachy tag should be removed?

**Comment by srish** [comment]
@misaochan Sounds good! removed :)

**Comment by gupta-meghna64** [comment]
@srish do we need to submit our phabricator task as well or just creating the task should be fine for the proposal deadline?

**Comment by misaochan** [comment]
I just received an email from Sage Sharp regarding the next round of Outreachy internships, Dec 18 - Mar 19. Unfortunately I personally cannot mentor this round as I will be unavailable for some periods of time, but is anyone interested in doing so? Please feel free to create a project etc if you are, thanks!

**Comment by nicolas-raoul** [comment]
I will be unavailable during the whole first half, so unfortunately I can't make it.
But I hope some of the people who are very active here can step in :-)
